Spec/RealtimeClientChannel.swift

```diff
@@ -1118,6 +1111,76 @@ class RealtimeClientChannel: QuickSpec {
                     }
                 }
 
+                // RTL5i
+                it("if the channel is in a pending state DETACHING, do the detach operation after the completion of the pending request") {
+                    let options = AblyTests.commonAppSetup()
+                    let client = ARTRealtime(options: options)
+                    defer { client.dispose(); client.close() }
+                    let channel = client.channels.get("foo")
+                    waitUntil(timeout: testTimeout) { done in
+                        channel.attach() { error in
+                            expect(error).to(beNil())
+                            done()
+                        }
+                    }
+
+                    var detachedCount = 1
+                    channel.on(.Detached) { stateChange in
+                        expect(stateChange?.reason).to(beNil())
+                        detachedCount += 1
+                    }
+
+                    waitUntil(timeout: testTimeout) { done in
+                        channel.once(.Detaching) { stateChange in
+                            guard let stateChange = stateChange else {
+                                fail("ChannelStateChange is nil"); done(); return
+                            }
+                            expect(stateChange.reason).to(beNil())
+                            expect(stateChange.current).to(equal(ARTRealtimeChannelState.Detaching))
+                            expect(stateChange.previous).to(equal(ARTRealtimeChannelState.Attached))
+                            channel.detach()
+                            done()
+                        }
+                        channel.detach()
+                    }
+
+                    expect(detachedCount) == 1
```

Comment:
This is racy, you don't allow enough time for the channel to potentially move into other states as you complete the test as soon as it reaches the detaching state which is triggered by the `channel.detach()`.  I would at a minimum add a timer to avoid this race and also check how many DETACHING states you see.